### ISSUE TITLE
Set of Metrics and Loss functions

### ISSUE THREAD

**Issue by SebastianJiroSchlecht** [artificial-audio/pyFDN#178]
Gallery function with
- similarity (h1, h2)
- metrics (h)
- metrics (on FDN parameter; or on modal representation)

such as
- EDR
- MR-STFT
- Mel Scale
- Spectral Flatness
- Roughness
- Echo density
- long autocorrelation as repetition measure (new idea)
...

**Comment by jeremybbq** [comment]
import auraloss or torchaudio?

**Comment by SebastianJiroSchlecht** [comment]
I would prefer that you reimplement those losses rather than add new dependencies. At least at first glance, torchaudio is huge, and we need only a fraction, right? Auroloss might be a good start to get things going (if the API is suitable).

I believe more losses will be added over time, so a consistent implementation is helpful.

**Comment by SebastianJiroSchlecht** [comment]
There should be gallery functions for
- features: RIR -> feature representation (e.g time-domain -> STFT; time-domain -> echo density, eigenvalue of MIMO RIR per frequency)
- loss: L(features 1, feature 2) ; e.g., L1, L2, ...

If possible, all should differentiable and in pyTorch.

Test with FLAMO training (see example example_train_colorless_FDN.py).

Extend the list of losses and features. Look at recent papers on FDNs. 

Also look at Auraloss library; also check FLAMO library.

**Comment by lsbharadwaj** [comment]
Dear @jeremybbq, @SebastianJiroSchlecht,

Kindly provide your input. I found the following list of loss function and features from the internet. But I am not sure which of these are of importance to you. Kindly find the following below
1. Table of Loss functions
2. List of lost functions **(Kindly tick the ones that are to be implemented)**
3. Table of Features and Loss functions that use the feature
4. List of Features **(Kindly tick the ones that are to be imiplemented)**


Kindly mark it by clicking on the checkboxes in the list below

### Loss function summary
S.No | Loss function | Features required for computation
-- | -- | --
1 | L1 | Any feature
2 | L2 / MSE | Any feature
3 | L∞ | Any feature
4 | Log-spectral distance | STFT magnitude
5 | Spectral convergence | STFT magnitude / MR-STFT
6 | STFT loss | STFT magnitude / phase
7 | Mel-spectrogram MSE | Mel spectrogram
8 | Multi-resolution STFT loss | MR-STFT
9 | Random-resolution STFT loss | STFT
10 | Multi-scale spectral loss | Multi-scale STFT / spectral representations
11 | Perceptual audio loss | Waveform / STFT / Mel
12 | ESR loss | Waveform
13 | DC loss | Waveform
14 | Log-cosh loss | Waveform
15 | SNR loss | Waveform
16 | SI-SDR loss | Waveform
17 | SD-SDR loss | Waveform
18 | EDR loss | Energy Decay Relief
19 | Energy-decay loss | Energy Decay Curve (EDC)
20 | Echo-density loss | Echo density
21 | Spectral-flatness loss | Spectral flatness
22 | Roughness loss | Roughness
23 | Autocorrelation loss | Autocorrelation
24 | Long-lag autocorrelation loss | Long-lag autocorrelation
25 | Spectral-centroid loss | Spectral centroid
26 | Modal matching loss | MIMO RIR eigenvalues / modal representation
27 | Modal-density loss | Modal density

### Loss functions checkbox

- [x] L1
- [x] L2 / MSE
- [ ] L∞
- [ ] Log-spectral distance
- [ ] Spectral convergence
- [ ] STFT loss
- [ ] Mel-spectrogram MSE
- [ ] Multi-resolution STFT loss
- [ ] Random-resolution STFT loss
- [ ] Multi-scale spectral loss
- [ ] Perceptual audio loss
- [x] ESR loss
- [x] DC loss
- [x] Log-cosh loss
- [x] SNR loss
- [x] SI-SDR loss
- [x] SD-SDR loss
- [x] EDR loss
- [ ] Energy-decay loss
- [ ] Echo-density loss
- [ ] Spectral-flatness loss
- [ ] Roughness loss
- [ ] Autocorrelation loss
- [ ] Long-lag autocorrelation loss
- [ ] Spectral-centroid loss
- [ ] Modal matching loss
- [ ] Modal-density loss

### Features summary

S.No | Feature | Loss functions
-- | -- | --
1 | STFT magnitude | L1, L2/MSE, L∞, Log-spectral distance, Spectral convergence, STFT loss, MR-STFT loss, Multi-scale spectral loss
2 | STFT phase | L1, L2/MSE, STFT loss
3 | MR-STFT | L1, L2/MSE, Spectral convergence, MR-STFT loss, Multi-scale spectral loss
4 | Mel spectrogram | L1, L2/MSE, Mel-spectrogram MSE, Perceptual audio loss
5 | Echo density | L1, L2/MSE, Echo-density loss
6 | MIMO RIR eigenvalues | L1, L2/MSE, Modal matching loss
7 | Spectral centroid | L1, L2/MSE, Spectral-centroid loss
8 | Energy decay curve (EDC) | L1, L2/MSE, Energy-decay loss
9 | Energy Decay Relief (EDR) | L1, L2/MSE, EDR loss
10 | Spectral flatness | L1, L2/MSE, Spectral-flatness loss
11 | Roughness | L1, L2/MSE, Roughness loss
12 | Autocorrelation | L1, L2/MSE, Autocorrelation loss
13 | Long-lag autocorrelation | L1, L2/MSE, Long-lag autocorrelation loss
14 | Modal density | L1, L2/MSE, Modal-density loss
15 | Modal frequencies | L1, L2/MSE, Modal matching loss
16 | Modal decay rates | L1, L2/MSE, Modal matching loss
17 | FDN delay-line distribution | L1, L2/MSE
18 | FDN delay ratios | L1, L2/MSE
19 | Feedback-matrix eigenvalues | L1, L2/MSE, Modal matching loss
20 | Feedback-matrix spectral radius | L1, L2/MSE
21 | Feedback-matrix mixing properties | L1, L2/MSE, Matrix-distance loss

### Features checkbox

- [x] STFT magnitude
- [x] STFT phase
- [ ] MR-STFT
- [x] Mel spectrogram
- [ ] Echo density
- [x] MIMO RIR eigenvalues
- [ ] Spectral centroid
- [x] Energy decay curve (EDC)
- [x] Energy Decay Relief (EDR)
- [x] Spectral flatness
- [ ] Roughness
- [ ] Autocorrelation
- [ ] Long-lag autocorrelation
- [ ] Modal density
- [ ] Modal frequencies
- [ ] Modal decay rates
- [ ] FDN delay-line distribution
- [ ] FDN delay ratios
- [ ] Feedback-matrix eigenvalues
- [ ] Feedback-matrix spectral radius
- [ ] Feedback-matrix mixing properties

**Comment by lsbharadwaj** [comment]
hi @almahova ,
Kindly pull the branch 178-set-of-metrics-and-loss-functions if you want to make changes. I have so far added STFT feature and an example code.

**Comment by SebastianJiroSchlecht** [comment]
@lsbharadwaj 
Thanks for the list. Can you have a look at the current repo and how these losses are implemented?
